shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -59,26 +61,131 @@
  * need to be parsed out of individual messages. Notably this includes the event number of the last
  * event in the block, so that it can be acknowledged.
  *
+ * if the channel is not free, then each append is enqueued to the session buffer.
+ * This session buffer gets flushed under following conditions
+ *      if the session buffer exceed the threshold (MAX_BLOCK_SIZE)
+ *      if the multiple sessions are present , then total session buffer exceed the threshold (MAX_DATA_SIZE)
+ *      if the total number of events exceeds the threshold (MAX_EVENTS)
+ *
  */
 @NotThreadSafe
 @RequiredArgsConstructor
 @Slf4j
 public class CommandEncoder extends MessageToByteEncoder<Object> {
     private static final byte[] LENGTH_PLACEHOLDER = new byte[4];
+    private static final int MAX_DATA_SIZE = 4 * 1024 * 1024; // 4MB
     private final Function<Long, AppendBatchSizeTracker> appendTracker;
     private final Map<Map.Entry<String, UUID>, Session> setupSegments = new HashMap<>();
     private final AtomicLong tokenCounter = new AtomicLong(0);
     private String segmentBeingAppendedTo;
     private UUID writerIdPerformingAppends;
     private int currentBlockSize;
     private int bytesLeftInBlock;
+    private final Map<UUID, Session> pendingWrites = new HashMap<>();
+    private long pendingBytes = 0;
+
 
     @RequiredArgsConstructor
-    private static final class Session {
+    private final class Session {
+        private static final int MAX_BLOCK_SIZE = 1024 * 1024;  // 1MB
+        private static final int MAX_EVENTS = 500;
         private final UUID id;
-        private long lastEventNumber = -1L;
-        private int eventCount;
         private final long requestId;
+        private final ByteBuf buffer = buffer();
```

Comment:
We should the api  `io.netty.buffer.Unpooled#buffer(int, int)` so as to explicitly set the `maxCapacity`

Reply:
default minimum is 256 ; maximum is 1MB which taken care by flush method. 
we limit at the buffer allocation level, then it requires additional code take care of buffer border conditions.  The allocated which more than 1MB , gets cleared after the flush or flushAll .
The flushAll gets triggered for every 100ms; so in this design, the buffer accumulation is can happen for at max 100ms only.

Reply:
yes.. its done now..